src/embedding/kd.ts

```diff
@@ -0,0 +1,752 @@
+     *    be posted
+     *
+     * @remarks use {@link DeleteDashboardCommand} factory function to instantiate
+     */
+    export type DeleteDashboardCommand = IKdMessageEvent<GdcKdCommandType.Delete, DeleteDashboardBody>;
+
+    export type DeleteDashboardCommandData = IKdGdcMessageEnvelope<
+        GdcKdCommandType.Delete,
+        DeleteDashboardBody
+    >;
+
+    /**
+     * Type-guard checking whether object is an instance of {@link DeleteDashboardData}.
+     *
+     * @param obj - object to test
+     */
+    export function isDeleteDashboardCommandData(obj: any): obj is DeleteDashboardCommandData {
+        return getEventType(obj) === GdcKdCommandType.Delete;
+    }
+
+    /**
+     * It's main content is empty.
+     */
+    export type SwitchToEditBody = IKdAvailableCommands;
+
+    /**
+     * Switches current dashboard to edit mode.
+     *
+     * Contract:
+     *
+     * -  if KD shows dashboard in view mode, will switch to edit mode and post SwitchedToEdit once ready for
+     *    editing
+     * -  if KD shows dashboard in edit mode, will keep edit mode and post SwitchedToEdit as if just switched
+     *    from view mode
+     * -  if no dashboard currently displayed, posts KdCommandFailed
+     *
+     * @remarks use {@link SwitchToEditCommand} factory function to instantiate
+     */
+    export type SwitchToEditCommand = IKdMessageEvent<GdcKdCommandType.SwitchToEdit, SwitchToEditBody>;
+
+    export type SwitchToEditCommandData = IKdGdcMessageEnvelope<
+        GdcKdCommandType.SwitchToEdit,
+        SwitchToEditBody
+    >;
+
+    /**
+     * Type-guard checking whether object is an instance of {@link isSwitchToEditCommandData}.
+     *
+     * @param obj - object to test
+     */
+    export function isSwitchToEditCommandData(obj: any): obj is SwitchToEditCommandData {
+        return getEventType(obj) === GdcKdCommandType.SwitchToEdit;
+    }
+
+    /**
+     * It's main content is empty.
+     */
+    export type ExportBody = IKdAvailableCommands;
+
+    /**
+     * Exports dashboard to PDF.
+     *
+     * Contract:
+     *
+     * -  if KD shows dashboard in view mode, will export dashboard to PDF and post ExportFinished once ready for
+     *    exporting
+     * -  if KD shwows dashboard in edit mode or not not showing any dashboard, KdCommandFailed will
+     *    be posted
+     * @remarks use {@link ExportCommand} factory function to instantiate
+     */
+    export type ExportCommand = IKdMessageEvent<GdcKdCommandType.Export, ExportBody>;
+
+    export type ExportCommandData = IKdGdcMessageEnvelope<GdcKdCommandType.SwitchToEdit, SwitchToEditBody>;
+
+    /**
+     * Type-guard checking whether object is an instance of {@link isExportCommandData}.
+     *
+     * @param obj - object to test
+     */
+    export function isExportCommandData(obj: any): obj is ExportCommandData {
+        return getEventType(obj) === GdcKdCommandType.Export;
+    }
+
+    export type IAddfilterBody = IKdAvailableCommands;
```

Comment:
body of adding filter command cannot be empty, can it?
It should be kind of attr uri/identifier
For now, I suggest not to add type for both addFilter and addWidget, we can do it next phase

Reply:
Remove it, will handle into next phase